// @suc/monch/src/utils/proj.js
import {Projection, addProjection, addCoordinateTransforms} from 'ol/proj.js';
import {applyTransform} from 'ol/extent';
import * as projzh from 'projzh'

function smerc2gcj(input, opt_output, opt_dimension) {
    let output = projzh.projection.sphericalMercator.inverse(input, opt_output, opt_dimension);
    output = projzh.datum.gcj02.fromWGS84(output, output, opt_dimension);
    return projzh.projection.sphericalMercator.forward(output, output, opt_dimension);
}

function gcj2smerc(input, opt_output, opt_dimension) {
    let output = projzh.projection.sphericalMercator.inverse(input, opt_output, opt_dimension);
    output = projzh.datum.gcj02.toWGS84(output, output, opt_dimension);
    return projzh.projection.sphericalMercator.forward(output, output, opt_dimension);
}

function ll2gcj(input, opt_output, opt_dimension) {
    let output = projzh.datum.gcj02.fromWGS84(input, opt_output, opt_dimension);
    return projzh.projection.sphericalMercator.forward(output, output, opt_dimension);
}

function gcj2ll(input, opt_output, opt_dimension) {
    let output = projzh.projection.sphericalMercator.inverse(input, opt_output, opt_dimension);
    return projzh.datum.gcj02.toWGS84(output, output, opt_dimension);
}

// 添加 gcj02 坐标系
export function addGcj02Projection() {
    const extent = [-180, -90, 180, 90];
    const gcj02 = new Projection({
        code: 'GCJ02',
        extent: applyTransform(extent, ll2gcj),
        units: 'm'
    });

    addProjection(gcj02);
    addCoordinateTransforms('EPSG:4326', gcj02, ll2gcj, gcj2ll);
    addCoordinateTransforms('EPSG:3857', gcj02, smerc2gcj, gcj2smerc);
}

// 添加百度坐标系
export function addBdProjection() {
    const extent = [72.004, 0.8293, 137.8347, 55.8271];
    const baiduMercator = new Projection({
        code: 'baidu',
        extent: applyTransform(extent, projzh.ll2bmerc),
        units: 'm'
    });

    addProjection(baiduMercator);
    addCoordinateTransforms('EPSG:4326', baiduMercator, projzh.ll2bmerc, projzh.bmerc2ll);
    addCoordinateTransforms('EPSG:3857', baiduMercator, projzh.smerc2bmerc, projzh.bmerc2smerc);
}
